### PR TITLE
Fused MLA override

### PR DESCRIPTION
**Pull request by SherlockNoMad**
# Fused DeepSeek-V3 MLA

## Summary

Adds an opt-in fused MLA implementation for TorchTitan's DeepSeek-V3 attention.

The override:

- Applies Q RoPE in place, avoiding Q concatenation.
- Fuses K RoPE, head expansion, and K materialization.
- Keeps V as a zero-copy view.
- Fuses KV-gradient packing, positional-gradient reduction, and inverse RoPE.
- Exposes the kernels as `torch.library` custom operators so they remain visible
  to GraphTrainer and `make_fx`.
- Uses 64-bit indexing for production DeepSeek-V3 tensor sizes.

The core fusion strategy is inspired by Megatron Core's fused MLA design. This
is an independent TorchTitan implementation using BSHD layouts, `ComplexRoPE`,
explicit position IDs, and GraphTrainer-compatible custom operators; no
Megatron Core or Transformer Engine kernel code is copied.

## Performance

Measured on one NVIDIA GB300 with BF16 tensors, `B=16`, `L=4096`, `H=128`.

### MLA assembly microbenchmark

The activation-checkpointing workload is two forwards plus one backward
(`2F + 1B`).

| Operation | Stock eager | Tuned fused | Reduction |
|---|---:|---:|---:|
| Forward | 19.160 ms | 1.276 ms | 93.3% |
| Backward | 15.579 ms | 1.949 ms | 87.5% |
| `2F + 1B` | 53.898 ms | 4.501 ms | **91.6%** |

The latest head-tiling tuning also makes the fused boundary **3.82x faster**
than the initial fused implementation.

### DeepSeek-V3 671B GraphTrainer

Configuration:

- Virtual FSDP degree: 256
- Expert parallel degree: 64
- Local/global batch: 16 / 4096
- Sequence length: 4096
- Gradient accumulation: 1
- GraphTrainer fake communication backend
- AOT FX graph with selective rematerialization and FSDP overlap scheduling

| Metric | Stock | Fused MLA | Change |
|---|---:|---:|---:|
| Throughput | 2,060 tok/s | 2,277 tok/s | **+10.53%** |
| Model TFLOP/s | 579.30 | 640.32 | **+10.53%** |
| MFU | 23.17% | 25.61% | +2.44 pp |
| GPU kernel span | 31.658 s | 28.805 s | **-9.01%** |
| Active-step memory | 182.46 GiB | 180.55 GiB | -1.91 GiB |

Both graphs retain the same 361 FSDP overlap dependencies.

In the profiled window, the targeted 4-D MLA `CatArrayBatchedCopy` kernel
accounts for 309 launches and 2.601 seconds in the stock run. The fused run
removes all of those launches; its three MLA kernels consume approximately
273 ms in total.

Overall CUDA kernel time falls from 29.954 seconds to 26.906 seconds.

## Validation

```bash
pytest -q tests/unit_tests/test_fused_mla_override.py
```

Result: **10 passed**.

Coverage includes:

- BF16 and FP32 forward/backward parity
- Exact copying of Q-NoPE, K-NoPE, V, and packed KV gradients
- Module-level attention parity
- Storage and aliasing behavior
- Position broadcasting
- State-dict compatibility
- Custom-op visibility under fake tensors and `make_fx`

The full trainer's step-1 loss and gradient norm match the stock implementation.

Formatting, lint, spelling, documentation, type, and Python compilation checks
also pass.

## Scope and limitations

- The override currently supports DeepSeek-V3 MLA with `ComplexRoPE`.
- `CosSinRoPE`, `MRoPE`, and non-MLA attention modules are not supported.
- The E2E benchmark uses one physical GB300 with fake 256-rank communication.
  It measures full local graph execution and scheduling, but not real multi-node
  networking or communication contention.
- CUDA graph conversion was skipped for both compared variants.

### CODE REVIEW

File: torchtitan/overrides/fused_mla.py [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyrefly: ignore-errors
+
+"""Opt-in fused DeepSeek-V3 MLA Q/KV assembly.
+
+Activate with::
+
+    --override.imports torchtitan.overrides.fused_mla.fused_mla
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_mla.py:13]: could you add this to the existing deepseek CI
https://github.com/pytorch/torchtitan/blob/main/tests/integration_tests/models.py#L100

or if you want to guard it on cudagraph, need to create a new one without EP

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/overrides/fused_mla.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/overrides/fused_mla.py:13]: added test with override. I am not too worry about the cudagraph.

File: tests/integration_tests/models.py [hunk 2]
```diff
@@ -107,6 +107,21 @@ def build_model_tests_list() -> list[OverrideDefinitions]:
...
             ngpu=4,
         ),
+        OverrideDefinitions(
+            [
+                [
+                    "--module deepseek_v3 --config deepseek_v3_debugmodel_mtp",
+                    "--override.imports torchtitan.overrides.fused_mla.fused_mla,"
+                    "torchtitan.overrides.fused_swiglu.fused_swiglu",
+                    "--parallelism.data_parallel_shard_degree 4",
+                    "--parallelism.expert_parallel_degree 2",
```
Comment (tianyu-l) [inline tests/integration_tests/models.py:118]: default EP backend won't work with cudagraph which is by default turned on

**Review by sanketpurandare — APPROVED** [review]
1. Add cudagraphable integration test with minimal async ep instead of default ep backend
2. tuned configs are great but you can pull all the kernel configs in a dataclass with defaults set to the tuned ones that would be great, would help others using MLA. Also ask the agent to insert a one line comment explaining the config.

File: tests/integration_tests/models.py (hunk 2 above)
Comment (SherlockNoMad) [reply tests/integration_tests/models.py:118]: good catch.

Comment (SherlockNoMad) [reply tests/integration_tests/models.py:118]: changed to MinimalAsyncEP so it's cudagraphable.

Comment (SherlockNoMad) [reply tests/integration_tests/models.py:118]: ah... A100 Ci machine doesn't support MinimalAsyncEP, so I am reverting back and disable cudagraph.
